runtime/drivers/azure/azure.go

```diff
@@ -0,0 +1,363 @@
+	}
+	bucketURL, err := globutil.ParseBucketURL(conf.Path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse path %q, %w", conf.Path, err)
+	}
+	if bucketURL.Scheme != "azblob" {
+		return nil, fmt.Errorf("invalid scheme %q in path %q", bucketURL.Scheme, conf.Path)
+	}
+
+	conf.url = bucketURL
+	return conf, nil
+}
+
+// getClient returns a new azure blob client.
+func (c *Connection) getClient(ctx context.Context, conf *sourceProperties) (*container.Client, error) {
+	var client *container.Client
+	var accountName, accountKey, sasToken, connectionString string
+	azClientOpts := &container.ClientOptions{}
+
+	if c.config.AllowHostAccess {
+		accountName = os.Getenv("AZURE_STORAGE_ACCOUNT")
+		accountKey = os.Getenv("AZURE_STORAGE_KEY")
+		sasToken = os.Getenv("AZURE_STORAGE_SAS_TOKEN")
+		connectionString = os.Getenv("AZURE_STORAGE_CONNECTION_STRING")
+	} else {
+		accountName = c.config.Account
+		accountKey = c.config.Key
+		sasToken = c.config.SASToken
+		// Added it as per azure blob client implementation, can remove if not required
+		connectionString = c.config.ConnectionString
+	}
+
+	if accountName != "" {
+		svcURL := fmt.Sprintf("https://%s.blob.core.windows.net", accountName)
+		containerURL, err := url.JoinPath(svcURL, conf.url.Host)
+		if err != nil {
+			return nil, err
+		}
+		if accountKey != "" {
+			sharedKeyCred, err := azblob.NewSharedKeyCredential(accountName, accountKey)
+			if err != nil {
+				return nil, fmt.Errorf("failed azblob.NewSharedKeyCredential: %w", err)
+			}
+
+			client, err = container.NewClientWithSharedKeyCredential(containerURL, sharedKeyCred, azClientOpts)
+			if err != nil {
+				return nil, fmt.Errorf("failed container.NewClientWithSharedKeyCredential: %w", err)
+			}
+		} else if sasToken != "" {
+			serviceURL, err := azureblob.NewServiceURL(&azureblob.ServiceURLOptions{
+				AccountName: accountName,
+				SASToken:    sasToken,
+			})
+			if err != nil {
+				return nil, err
+			}
+
+			containerURL, err := url.JoinPath(string(serviceURL), conf.url.Host)
+			if err != nil {
+				return nil, err
+			}
+
+			client, err = container.NewClientWithNoCredential(containerURL, azClientOpts)
+			if err != nil {
+				return nil, fmt.Errorf("failed container.NewClientWithNoCredential: %w", err)
+			}
+		} else {
+			cred, err := azidentity.NewDefaultAzureCredential(nil)
+			if err != nil {
+				return nil, fmt.Errorf("failed azidentity.NewDefaultAzureCredential: %w", err)
+			}
+			client, err = container.NewClient(containerURL, cred, azClientOpts)
+			if err != nil {
+				return nil, fmt.Errorf("failed container.NewClient: %w", err)
+			}
+		}
+	} else if connectionString != "" {
+		var err error
+		client, err = container.NewClientFromConnectionString(connectionString, conf.url.Host, azClientOpts)
+		if err != nil {
+			return nil, fmt.Errorf("failed container.NewClientFromConnectionString: %w", err)
+		}
+	} else {
+		return nil, errors.New("internal error, no account name")
```

Comment:
This is not an internal error. Need to throw `drivers.NewPermissionDeniedError`.